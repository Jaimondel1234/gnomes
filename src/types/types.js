export const types = {
  login: "[Auth] Login",
  logout: "[Auth] Logout",
  loginRequest: "[Auth] Login Request",
};
