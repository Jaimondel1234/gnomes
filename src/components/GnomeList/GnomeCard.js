import React from 'react';
import {
  PanelStyle,
  ImageStyle,
  FlexStyle,
  ContainerStyle,
  ButtonStyle,
  SpanStyle,
} from '../../styles/core';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

/*
  Component that displays the gnome card
*/
export const GnomeCard = ({ ...props }) => {
  const { name, thumbnail } = props;
  const history = useHistory();
  const handleClick = () => {
    history.push('/gnome-detail', { ...props });
  };

  return (
    <PanelStyle
      mr='30px'
      mb='50px'
      maxWidth='900px'
      borderRadius='10px'
      pt='20px'
      mt='10px'
    >
      <FlexStyle
        alignItems='center'
        justifyContent='space-between'
        flexDirection='column'
      >
        <SpanStyle
          color='white'
          fontSize='30px'
          fontWeight='bold'
          p='15px'
          textShadow='2px 2px 4px #000000'
          borderRadius='5px'
          bg='#1E90FF'
        >
          {name}
        </SpanStyle>
        <ContainerStyle p='20px'>
          <ImageStyle
            src={thumbnail}
            width='350px'
            height='300px'
            onError={(e) => {
              e.target.onerror = null;
              e.target.src =
                'http://www.publicdomainpictures.net/pictures/30000/nahled/maple-leaves-background.jpg';
            }}
          />
        </ContainerStyle>
        <ButtonStyle
          primary
          p='12px'
          fontSize='25px'
          fontWeight='bold'
          onClick={handleClick}
          letterSpacing='3px'
          lineHeight='50px'
        >
          ... More
        </ButtonStyle>
      </FlexStyle>
    </PanelStyle>
  );
};

GnomeCard.propTypes = {
  props: PropTypes.shape({
    name: PropTypes.string,
    thumbnail: PropTypes.string,
    age: PropTypes.number,
    genre: PropTypes.string,
    hair_color: PropTypes.string,
    height: PropTypes.number,
    weight: PropTypes.number,
    professions: PropTypes.arrayOf(PropTypes.string),
    friends: PropTypes.arrayOf(PropTypes.string),
  }),
};
