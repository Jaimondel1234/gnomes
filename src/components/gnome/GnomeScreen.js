import React from "react";

export const GnomeScreen = () => {
  return (
    <div>
      <h1>Gnome Screen</h1>
    </div>
  );
};
