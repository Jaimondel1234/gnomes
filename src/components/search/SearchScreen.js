import React from 'react';
import { ContainerStyle, TextStyle } from '../../styles/core';
import { SearchInput } from './SearchInput';
import { useSelector } from 'react-redux';
import { GnomeInformation } from '../Shared/GnomeInformation';
import { isEmpty } from 'lodash';
/*
component that displays the search screen for a gnome
*/
export const SearchScreen = () => {
  const { gnomeSearched } = useSelector((state) => state.gnome);
  return (
    <ContainerStyle m='20px'>
      <TextStyle fontSize='30px' fontWeight='bold' mb='10px'>
        Search the gnome:
      </TextStyle>
      <SearchInput />
      {!isEmpty(gnomeSearched) && <GnomeInformation {...gnomeSearched} />}
    </ContainerStyle>
  );
};
