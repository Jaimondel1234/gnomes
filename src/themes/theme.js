/* 
  Theme
*/

export const theme = {
  color: {
    main: '#007bff',
  },
};
