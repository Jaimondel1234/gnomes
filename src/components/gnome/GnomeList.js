import React, { useRef, useEffect } from "react";
import { FlexStyle, ScrollContainer } from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <>
      <ScrollContainer ref={ref} onScroll={handleScroll}>
        <FlexStyle
          alignItem="center"
          justifyContent="center"
          flexDirection="row"
          flexWrap="wrap"
        >
          {gnomes &&
            gnomes.map(({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        </FlexStyle>
      </ScrollContainer>
    </>
  );
};
