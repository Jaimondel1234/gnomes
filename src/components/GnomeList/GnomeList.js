import React, { useRef, useState, useEffect } from "react";
import {
  FlexStyle,
  ScrollContainer,
  PlaceHolderStyle,
} from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";
import { fill } from "lodash";
import map from "lodash/map";

let emptyArray = fill(new Array(20), 0);

const placeHolderArray = emptyArray.map((value, id) => {
  return `place_holder_${id}`;
});

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes, loading } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <ScrollContainer ref={ref} onScroll={handleScroll}>
      <FlexStyle
        alignItem="center"
        justifyContent="center"
        flexDirection="row"
        flexWrap="wrap"
      >
        {gnomes &&
          map(gnomes, ({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        {loading &&
          map(placeHolderArray, (value) => (
            <PlaceHolderStyle
              key={value}
              bg="grey"
              mr="30px"
              mb="50px"
              maxWidth="900px"
              borderRadius={10}
              pt="50px"
            />
          ))}
      </FlexStyle>
    </ScrollContainer>
  );
};
