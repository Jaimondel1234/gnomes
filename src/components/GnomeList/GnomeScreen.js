import React, { useEffect } from "react";
import { ContainerStyle, ScrollContainer } from "../../styles/core";
import { GnomeList } from "./GnomeList";
import { loadGnomes } from "../../actions/gnome";
import { useDispatch, useSelector } from "react-redux";

export const GnomeScreen = () => {
  const dispatch = useDispatch();
  const { scrollNumberGnomes } = useSelector((state) => state.gnome);

  useEffect(() => {
    if (scrollNumberGnomes === 0) {
      dispatch(loadGnomes());
    }
  }, [loadGnomes, dispatch, scrollNumberGnomes]);

  return (
    <ContainerStyle mt="5">
      <GnomeList />
    </ContainerStyle>
  );
};
