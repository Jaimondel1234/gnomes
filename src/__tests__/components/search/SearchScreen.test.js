import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";

import { GnomeScreen } from "../../../components/GnomeList/GnomeScreen";
import { theme } from "../../../themes/theme";
import { types } from "../../../types/types";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { gnomesMock } from "../../../__mocks__/gnomesMock";

describe("Test <SearchScreen/>", () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const initState = {
    gnome: { gnomeSearched: gnomesMock[0] },
  };

  const store = mockStore(initState);
  store.dispatch = jest.fn();
  const wrapper = mount(
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <SearchScreen />
      </ThemeProvider>
    </Provider>
  );
  test("it must be displayed correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("if there is a search for a gnome, the gnome information must be displayed", () => {
    expect(wrapper.find("GnomeInformation").props("props")).toEqual(
      gnomesMock[0]
    );
  });
});
