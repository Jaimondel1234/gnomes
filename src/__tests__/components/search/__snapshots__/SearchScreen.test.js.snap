// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <SearchScreen/> it must be displayed correctly 1`] = `
<div
  className="core__ContainerStyle-qu5lcc-0 jcbrOh"
>
  <div
    className="core__TextStyle-qu5lcc-3 WnKye"
    fontSize="30px"
    fontWeight="bold"
  >
    Search the gnome:
  </div>
  <input
    autoComplete="off"
    className="core__InputStyle-qu5lcc-5 SearchInput__SearchInputStyle-uz7aty-0 eMTzQF GMxWa"
    name="searchText"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Tobus Quickwhistle"
    type="text"
    value=""
  />
  <div
    className="core__FlexStyle-qu5lcc-4 bwoVxe"
  >
    <span
      className="core__SpanStyle-qu5lcc-9 eYglRQ"
      color="white"
      fontSize="60px"
      fontWeight="bold"
    >
      Gnome Information
    </span>
    <div
      className="core__PanelStyle-qu5lcc-1 bKsISG"
      color="#ffff"
    >
      <span
        className="core__SpanStyle-qu5lcc-9 kHaOjz"
        color="white"
        fontSize="40px"
        fontWeight="bold"
      >
        Tobus Quickwhistle
      </span>
      <div
        className="core__FlexStyle-qu5lcc-4 ifNWpz"
      >
        <img
          className="core__ImageStyle-qu5lcc-7 keFsCu"
          onError={[Function]}
          src="http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg"
        />
        <ul
          className="core__ContainerList-qu5lcc-12 bqAzqZ"
        >
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 dOBnnW eYDWhd"
          >
            Age: 
            306
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Genre: 
            female
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Hair color: 
            Pink
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Height: 
            107.75835
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Width: 
            39.065952
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Professions:
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Metalworker
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Woodcarver
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Stonecarver
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
             Tinker
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Tailor
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Potter
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 ffJUNN eYDWhd"
          >
            Professions:
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Cogwitz Chillwidget
          </li>
          <li
            className="core__ListElement-qu5lcc-13 GnomeInformation__ListElementInformation-xcmbfp-0 GnomeInformation__SquareListElementInformation-xcmbfp-1 ffJUNN eYDWhd hFomdp"
          >
            Tinadette Chillbuster
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
